README.md
<p align="center">
    <img src="./sources/logo.png" alt="Image" width="320" height="304"/>
</p>

# markdown-it-graphviz（基于 markdown-it 的 graphviz 扩展）

![nodejs>8.0](https://img.shields.io/badge/nodejs-%3E%208.0-brightgreen.svg?id=markdown-it-graphviz)
![npm>6.0](https://img.shields.io/badge/npm-%20%3E%206.0-brightgreen.svg?id=markdown-it-graphviz)
![markdown_it_graphviz](https://img.shields.io/badge/markdown_it_graphviz-1.0.0-brightgreen.svg?id=markdown-it-graphviz)
![aiyou-bud](https://img.shields.io/badge/aiyou-bud-brightgreen.svg?id=markdown-it-graphviz)

- [🔥 功能列表](#功能列表)
- [✨ 使用介绍](#使用介绍)
- [🚀 快速上手](#快速上手)
- [📚 开发文档](#开发文档)
- [📞 社区交流](#社区交流)

---

## markdown-it-graphviz 是什么鬼？

markdown-it-graphviz 是由哎呦迪奥编写的基于 graphviz 在 web 页面中生成各种图形的**Markdown-IT**扩展插件，种类繁多，我主要是用来生成思维导图，非常方便使用。

- ✂️ 轻量：在不依赖 graphviz 和修改后的 viz.js 的基础上，只有 2kb。mermaid 其实也差不多，不管它。
- ⚡ 易用：你怎么使用 markdown-it 的插件，就怎么使用它。
- ⚙️️ 简单: 虽然有两种使用方式，但非常简单，以使用 markdown-it 插件的方式来使用它。

| nodejs | npm   | size |
| ------ | ----- | ---- |
| > 8.0  | > 6.0 | 2kb  |

## 🔥 功能列表

markdown-it-graphviz 插件，将 graphviz 语法转换成各种图形，也算功能强大。
graphviz 有很多种图的（流程图、时序图、思维导图、行为图等等等），你可以去[graphviz官网](http://www.graphviz.org/documentation/)看哟。
也许官网上看到的不太直观，你可以去网上搜索，然后学习，也可以参考我从网上搜索到的几篇文章。

- [graphviz的基本语法及使用](https://blog.csdn.net/mouday/article/details/80902992?depth_1-utm_source=distribute.pc_relevant.none-task&utm_source=distribute.pc_relevant.none-task)
- [Graphviz 画图的一些总结](https://www.cnblogs.com/shuqin/p/11897207.html)
- [Graphviz](https://www.jianshu.com/p/79f1a4acd2f2)

关于Graphviz语法代码高亮，可以在VS Code的扩展中安装插件，直接搜索 Graphviz，然后装装装，最后看看插件的使用教程。

---

## ✨ 使用介绍

### ⚔️ 安装指南

作为第三方依赖安装：

```bash
    npm install -S markdown-it-graphviz
```

### 📈 使用方法

---

局部使用（作为第三方依赖包来使用），在 node 环境下的 js 文件中复制以下代码进行粘贴：

```javascript
import markdownIt from 'markdown-it';
import markdownItGraphviz from 'markdown-it-graphviz';

// #region commonjs方式 Code Module

// const markdownItGraphviz = require('markdown-it-graphviz');

// #endregion commonjs方式 Code Module End

const mdi = markdownIt();
mdi.use(markdownItGraphviz);
mdi.render(`\`\`\`graphviz
digraph g {
	rankdir=LR  //方向左右
	dot语言->{简介,语法,示例}
	dot语言[shape=box,fontcolor=red]
	简介[color=red]
	语法[color=green]
	示例[color=blue]
	简介->{开源免费,UML绘图,导出svg}
	语法->{"digraph","graph"}
	"digraph"->导向图[label=可以制作带方向的导图]
	"graph"->无向图[label=可以制作不带方向的导图]
}
\`\`\``);
```

---

### 😎 效果

![思维导图](./sources/20200308071434.png)


---

## 🚀 快速上手

**安装**->**使用**

markdown-it 系列的插件怎么使用的，markdown-it-graphviz 就怎么使用。

可以参照我在维护的这个网站效果噢 [我的网站](http://www.hao6.website:1000/#graphviz-%E6%80%9D%E7%BB%B4%E5%AF%BC%E5%9B%BE)

## 📚 开发文档

安装node后，使用命令 **npm start** ，打开浏览器，
输入地址 http://localhost:8989/index.html 或者 http://127.0.0.1:8989
就可以看到效果了，markdown源文件是 sample.md，页面中展示的效果根据这个markdown文件来进行展示的。


## 代码仓库

[github 链接](https://github.com/aiyoudiao/markdown-it-graphviz)

## 📞 社区交流

markdown-it-graphviz 也算是**哎呦系列**中一员，虽然它没有以aiyou开头，之后还会有其它的成员陆续上来。

加群交流：722703370
![qq群](./sources/20200228230627.png)

### 感谢以下开源项目

- viz.js


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.2](https://github.com/aiyoudiao/markdown-it-graphviz/compare/v1.1.1...v1.1.2) (2020-03-09)


### Bug Fixes

* **common目录:** 修复了fs模块未能正常解析的bug ([bea178e](https://github.com/aiyoudiao/markdown-it-graphviz/commit/bea178ee9c7c99069c5f4e79d64610858f78f46f))

## [1.1.1](https://github.com/aiyoudiao/markdown-it-graphviz/compare/v1.1.0...v1.1.1) (2020-03-08)

### 📝 文档
范围|描述|commitId
--|--|--
 readme文档 | 增加预览效果的网站地址，这个网站我在维护哟，功能也是最新的。 | [520a746](https://github.com/aiyoudiao/markdown-it-graphviz/commit/520a746)


### chore
范围|描述|commitId
--|--|--
 release | 1.1.1 | [58471ed](https://github.com/aiyoudiao/markdown-it-graphviz/commit/58471ed)

# [1.1.0](https://github.com/aiyoudiao/markdown-it-graphviz/compare/4646744...v1.1.0) (2020-03-08)

范围|描述|commitId
--|--|--
 - | Initial commit | [4646744](https://github.com/aiyoudiao/markdown-it-graphviz/commit/4646744)


### 🌟 新功能
范围|描述|commitId
--|--|--
 整个项目 | markdown-it-graphviz的基本功能完成，可以正常使用 | [9071fce](https://github.com/aiyoudiao/markdown-it-graphviz/commit/9071fce)


### chore
范围|描述|commitId
--|--|--
 release | 1.1.0 | [19749cb](https://github.com/aiyoudiao/markdown-it-graphviz/commit/19749cb)


.eslintrc.js
/*
 * @Descripttion: ilovejwl
 * @version: 1.0.0
 * @Author: ilovejwl
 * @Date: 2020-01-18 18:24:21
 * @LastEditTime : 2020-02-04 09:06:51
 * @LastEditors  : ilovejwl
 */
module.exports = {
  root: true,
  parser: 'babel-eslint',
  extends: 'standard',
  plugins: ['html'],
  env: {
    // 环境
    browser: true, // 浏览器环境
    node: true, // node环境
    es6: true,
    commonjs: true,
  },
  globals: {
    // 设置全局变量
    $: true,
    layui: true,
    layer: true,
    okLoading: true,
    echarts: true,
    ClassicEditor: true,
    UE: true,
  },
  rules: {
    // 用来覆盖 extends : 'standrad'中标准的规则
    indent: ['error', 2], // indent 表示缩进的规则， error表示重要性，4表示使用4个空格缩进
    'eol-last': ['error', 'never'], // eol-last 表示最后一行要换行， never 表示不需要换行
    'func-call-spacing': ['error', 'always'],
    semi: ['error', 'always'],
    'prefer-const': [
      'error',
      {
        destructuring: 'any',
        ignoreReadBeforeAssign: false,
      },
    ],
    quotes: [
      'error',
      'single',
      {
        allowTemplateLiterals: true,
      },
    ],
    'no-tabs': ['error', {allowIndentationTabs: true}],
    'no-mixed-spaces-and-tabs': ['error', 'smart-tabs'],
    'no-new': 'off',
    'no-caller': 'off',
  },
};


common/viz.es-changed.js
/*
Viz.js 2.1.2 (Graphviz 2.40.1, Expat 2.2.5, Emscripten 1.37.36)
Copyright (c) 2014-2018 Michael Daines
Licensed under MIT license

This distribution contains other software in object code form:

Graphviz
Licensed under Eclipse Public License - v 1.0
http://www.graphviz.org

Expat
Copyright (c) 1998, 1999, 2000 Thai Open Source Software Center Ltd and Clark Cooper
Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006 Expat maintainers.
Licensed under MIT license
http://www.libexpat.org

zlib
Copyright (C) 1995-2013 Jean-loup Gailly and Mark Adler
http://www.zlib.net/zlib_license.html
*/




var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) {
  return typeof obj;
} : function (obj) {
  return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
};

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};

var ModuleWrapper = function ModuleWrapper(module, render) {
  classCallCheck(this, ModuleWrapper);

  var instance = module();
  this.render = function (src, options) {
    return render(instance, src, options)
  };
};

// https://developer.mozilla.org/en-US/docs/Web/API/WindowBase64/Base64_encoding_and_decoding


function b64EncodeUnicode(str) {
  return btoa(encodeURIComponent(str).replace(/%([0-9A-F]{2})/g, function (match, p1) {
    return String.fromCharCode('0x' + p1);
  }));
}

function defaultScale() {
  if ('devicePixelRatio' in window && window.devicePixelRatio > 1) {
    return window.devicePixelRatio;
  } else {
    return 1;
  }
}

function svgXmlToImageElement(svgXml) {
  var _ref = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref$scale = _ref.scale,
      scale = _ref$scale === undefined ? defaultScale() : _ref$scale,
      _ref$mimeType = _ref.mimeType,
      mimeType = _ref$mimeType === undefined ? "image/png" : _ref$mimeType,
      _ref$quality = _ref.quality,
      quality = _ref$quality === undefined ? 1 : _ref$quality;

  return new Promise(function (resolve, reject) {
    var svgImage = new Image();

    svgImage.onload = function () {
      var canvas = document.createElement('canvas');
      canvas.width = svgImage.width * scale;
      canvas.height = svgImage.height * scale;

      var context = canvas.getContext("2d");
      context.drawImage(svgImage, 0, 0, canvas.width, canvas.height);

      canvas.toBlob(function (blob) {
        var image = new Image();
        image.src = URL.createObjectURL(blob);
        image.width = svgImage.width;
        image.height = svgImage.height;

        resolve(image);
      }, mimeType, quality);
    };

    svgImage.onerror = function (e) {
      var error;

      if ('error' in e) {
        error = e.error;
      } else {
        error = new Error('Error loading SVG');
      }

      reject(error);
    };

    svgImage.src = 'data:image/svg+xml;base64,' + b64EncodeUnicode(svgXml);
  });
}

function svgXmlToImageElementFabric(svgXml) {
  var _ref2 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
      _ref2$scale = _ref2.scale,
      scale = _ref2$scale === undefined ? defaultScale() : _ref2$scale,
      _ref2$mimeType = _ref2.mimeType,
      mimeType = _ref2$mimeType === undefined ? 'image/png' : _ref2$mimeType,
      _ref2$quality = _ref2.quality,
      quality = _ref2$quality === undefined ? 1 : _ref2$quality;

  var multiplier = scale;

  var format = void 0;
  if (mimeType == 'image/jpeg') {
    format = 'jpeg';
  } else if (mimeType == 'image/png') {
    format = 'png';
  }

  return new Promise(function (resolve, reject) {
    fabric.loadSVGFromString(svgXml, function (objects, options) {
      // If there's something wrong with the SVG, Fabric may return an empty array of objects. Graphviz appears to give us at least one <g> element back even given an empty graph, so we will assume an error in this case.
      if (objects.length == 0) {
        reject(new Error('Error loading SVG with Fabric'));
      }

      var element = document.createElement("canvas");
      element.width = options.width;
      element.height = options.height;

      var canvas = new fabric.Canvas(element, { enableRetinaScaling: false });
      var obj = fabric.util.groupSVGElements(objects, options);
      canvas.add(obj).renderAll();

      var image = new Image();
      image.src = canvas.toDataURL({ format: format, multiplier: multiplier, quality: quality });
      image.width = options.width;
      image.height = options.height;

      resolve(image);
    });
  });
}

var Viz = function () {
  function Viz() {
    var _ref3 = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
    Module = _ref3.Module,
    render = _ref3.render;
    
    classCallCheck(this, Viz);
    if (typeof Module !== 'undefined' && typeof render !== 'undefined') {
      this.wrapper = new ModuleWrapper(Module, render);
    }else if (typeof Viz.Module !== 'undefined' && typeof Viz.render !== 'undefined') {
      this.wrapper = new ModuleWrapper(Viz.Module, Viz.render);
    } else {
      throw new Error('Must specify workerURL or worker option, Module and render options, or include one of full.render.js or lite.render.js after viz.js.');
    }
  }


  createClass(Viz, [{
    key: 'renderString',
    value: function renderString(src) {
      var _ref4 = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {},
          _ref4$format = _ref4.format,
          format = _ref4$format === undefined ? 'svg' : _ref4$format,
          _ref4$engine = _ref4.engine,
          engine = _ref4$engine === undefined ? 'dot' : _ref4$engine,
          _ref4$files = _ref4.files,
          files = _ref4$files === undefined ? [] : _ref4$files,
          _ref4$images = _ref4.images,
          images = _ref4$images === undefined ? [] : _ref4$images,
          _ref4$yInvert = _ref4.yInvert,
          yInvert = _ref4$yInvert === undefined ? false : _ref4$yInvert,
          _ref4$nop = _ref4.nop,
          nop = _ref4$nop === undefined ? 0 : _ref4$nop;

      for (var i = 0; i < images.length; i++) {
        files.push({
          path: images[i].path,
          data: '<?xml version="1.0" encoding="UTF-8" standalone="no"?>\n<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">\n<svg width="' + images[i].width + '" height="' + images[i].height + '"></svg>'
        });
      }

      return this.wrapper.render(src, { format: format, engine: engine, files: files, images: images, yInvert: yInvert, nop: nop });
    }
  }, {
    key: 'renderSVGElement',
    value: function renderSVGElement(src) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      return this.renderString(src, _extends({}, options, { format: 'svg' })).then(function (str) {
        var parser = new DOMParser();
        return parser.parseFromString(str, 'image/svg+xml').documentElement;
      });
    }
  }, {
    key: 'renderImageElement',
    value: function renderImageElement(src) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var scale = options.scale,
          mimeType = options.mimeType,
          quality = options.quality;


      return this.renderString(src, _extends({}, options, { format: 'svg' })).then(function (str) {
        if ((typeof fabric === 'undefined' ? 'undefined' : _typeof(fabric)) === "object" && fabric.loadSVGFromString) {
          return svgXmlToImageElementFabric(str, { scale: scale, mimeType: mimeType, quality: quality });
        } else {
          return svgXmlToImageElement(str, { scale: scale, mimeType: mimeType, quality: quality });
        }
      });
    }
  }, {
    key: 'renderJSONObject',
    value: function renderJSONObject(src) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      var format = options.format;


      if (format !== 'json' || format !== 'json0') {
        format = 'json';
      }

      return this.renderString(src, _extends({}, options, { format: format })).then(function (str) {
        return JSON.parse(str);
      });
    }
  }]);


  return Viz;
}();

export default Viz;
      

index.js

/*
 * @Descripttion: markdown-it-graphviz
 * @version: 1.0.0
 * @Author: aiyoudiao
 * @Date: 2020-03-07 04:43:27
 * @LastEditTime: 2020-03-08 09:21:51
 * @LastEditors: aiyoudiao
 * @FilePath: \markdown-it-graphviz\index.js
 */

// #region 方式一 Code Module

const Viz = require ('./common/viz.es-changed.js').default;
const { render, Module } = require ('./common/full.render-changed.js');

// #endregion 方式一 Code Module End
// ------------------------分割线用来分割不同的使用方式---------------------------
// #region 方式二 Code Module

// const Viz = window.Viz;

// #endregion 方式二 Code Module End

module.exports = function markdownItGraphViz (md, options = {}) {
  // console.log(`@md`, md);

  const temp = md.renderer.rules.fence.bind (md.renderer.rules);
  md.renderer.rules.fence = (tokens, idx, options, env, slf) => {
    try {
      const { content, info } = tokens[idx];
      if (info === 'graphviz') {
        // console.log(`@content`, content);

        // #region 方式一 Code Module

        var viz = new Viz ({ render, Module });

        // #endregion 方式一 Code Module End
        // ------------------------分割线用来分割不同的使用方式---------------------------
        // #region 方式二 Code Module

        // var viz = new Viz({});

        // #endregion 方式二 Code Module End

        const $svg = viz.wrapper.render (content, { format: 'svg', engine: 'dot', files: [], images: [], yInvert: false, nop: 0 });

        return $svg;
      }
    } catch (error) {
      return `<p style="border: 2px dashed red">Failed to render graphviz<span>${md.utils.escapeHtml (error.toString ())}</span></p>`;
    }
    return temp (tokens, idx, options, env, slf);
  };
};